### PR TITLE
Add customised mapping for tag1, tag2 and tag3

### LINKED ISSUE #89: Add support for the customisable tag selector
There's a tag selector we implemented in Magento 1.x and this is missing on M2. We should add a screen that allows the merchant to select the tag-1, tag-2 and tag-3 attributes.

### PR DESCRIPTION
**Pull request by chenliangde**
Closes #89

### CODE REVIEW

File: Helper/Data.php [hunk 1]
```diff
@@ -308,4 +323,37 @@ public function getPlatformVersion()
...
+     * @param StoreInterface $store the store model or null.
+     * @return array of attributes
+     */
+    public function getTag1Attributes(StoreInterface $store = null)
+    {
+        return $this->getStoreConfig(self::XML_PATH_TAG1, $store);
+    }
+
+    /**
+     * Get tag1 mapping attributes
```
Comment (mridang) [inline Helper/Data.php]: Typo

File: Helper/Data.php [hunk 2]
```diff
@@ -308,4 +323,37 @@ public function getPlatformVersion()
...
+     * @param StoreInterface $store the store model or null.
+     * @return array of attributes
+     */
+    public function getTag2Attributes(StoreInterface $store = null)
+    {
+        return $this->getStoreConfig(self::XML_PATH_TAG2, $store);
+    }
+
+    /**
+     * Get tag1 mapping attributes
```
Comment (mridang) [inline Helper/Data.php]: Typo

File: Helper/Data.php [hunk 3]
```diff
@@ -103,6 +103,21 @@ class Data extends AbstractHelper
     const XML_PATH_PRODUCT_UPDATES = 'nosto/flags/product_updates';
 
     /**
+     * Path to the configuration object for customized tag1
+     */
+    const XML_PATH_TAG1 = 'nosto/attributes/tag1';
+
+    /**
+     * Path to the configuration object for customized tag1
```
Comment (mridang) [inline Helper/Data.php]: Typo

File: Helper/Data.php [hunk 4]
```diff
@@ -103,6 +103,21 @@ class Data extends AbstractHelper
...
+     */
+    const XML_PATH_TAG2 = 'nosto/attributes/tag2';
+
+    /**
+     * Path to the configuration object for customized tag1
```
Comment (mridang) [inline Helper/Data.php]: Typo

File: Model/Product/Builder.php [hunk 5]
```diff
@@ -194,6 +196,9 @@ public function build(Product $product, Store $store)
             if (($tags = $this->buildTags($product)) !== []) {
                 $nostoProduct->setTag1($tags);
             }
+
+            //update customized tag1
```
Comment (mridang) [inline Model/Product/Builder.php]: Typo

File: Helper/Data.php [hunk 6]
```diff
@@ -308,4 +323,37 @@ public function getPlatformVersion()
...
         return $version;
     }
+
+    /**
+     * Get tag1 mapping attributes
+     *
+     * @param StoreInterface $store the store model or null.
+     * @return array of attributes
+     */
+    public function getTag1Attributes(StoreInterface $store = null)
```
Comment (hannunosto) [inline Helper/Data.php]: Do we really need separate methods for each getter? I'd go with M1 approach

File: Model/Config/Source/AttributeSelector.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+use Magento\Catalog\Model\ResourceModel\Product\Attribute\Collection;
+
+/**
+ * Option array class to generate a list of selectable options that allows the merchant to choose
+ * any attribute for his brand tag.
+ *
+ * @package Nosto\Tagging\Model\Config\Source
+ */
+class AttributeSelector extends Selector
```
Comment (hannunosto) [inline Model/Config/Source/AttributeSelector.php]: Rename class to something more descriptive (for example `Tags`) to follow the same naming convention as `Brand`, `Gtin`, etc.

File: Helper/Data.php [hunk 8]
```diff
@@ -103,6 +103,21 @@ class Data extends AbstractHelper
     const XML_PATH_PRODUCT_UPDATES = 'nosto/flags/product_updates';
 
     /**
+     * Path to the configuration object for customized tag1
+     */
+    const XML_PATH_TAG1 = 'nosto/attributes/tag1';
```
Comment (hannunosto) [inline Helper/Data.php]: Do we really need separate constants for each path? I'd go with M1 approach: https://github.com/Nosto/nosto-magento/blob/develop/app/code/community/Nosto/Tagging/Helper/Data.php#L50

File: Model/Product/Builder.php [hunk 9]
```diff
@@ -206,6 +211,41 @@ public function build(Product $product, Store $store)
...
+    {
+        foreach (self::CUSTOMIZED_TAGS as $tag) {
+            $getCustomizedAttribusMethodName = 'get' . $tag . 'Attributes';
+            $addTagMethodName = 'add' . $tag;
+            //getTag1Attributes(), getTag2Attributes() and getTag3Attributes() are called
+            $attributesConfig = $this->nostoDataHelper->$getCustomizedAttribusMethodName($store);
+            if ($attributesConfig == null) {
+                continue;
+            }
+            $attributes = explode(',', $attributesConfig);
```
Comment (chenliangde) [inline Model/Product/Builder.php]: Good point, fixed it.

**Pull request merged**